keg/component.py

```diff
@@ -0,0 +1,88 @@
+import importlib
+
+import flask
+
+
+class KegComponent:
+    """ Keg components follow the paradigm of flask extensions, and provide some defaults for the
+    purpose of setting up model/view structure. Using components, a project may be broken down into
+    logical blocks, each having their own entities, blueprints, templates, tests, etc.
+
+    Setup involves:
+    - KEG_REGISTERED_COMPONENTS config setting: assumed to be an iterable of importable dotted paths
+    - `__component__`: at the top level of each dotted path, this attribute should point to an
+    instance of `KegComponent`. E.g. `__component__ = KegComponent('widgets')`
+
+    By default, components will load entities from `db_visit_modules` into metadata and register
+    any blueprints specified by `blueprints`.
+
+    Blueprints can be created with the helper methods `get_named_blueprint` or `get_blueprint` in
+    order to have a configured template folder relative to the blueprint path.
+    """
+    db_visit_modules = ['.model.entities']
+    blueprints = [('.views', 'component_bp')]
+    template_folder = 'templates'
+
+    def __init__(self, name, app=None, db_visit_modules=None, blueprints=None,
+                 template_folder=None):
+        self.name = name
+        # This gets used as an absolute parent for the relative import paths of model/blueprints
+        self.relative_dotted_path = ''
+        # Allow customization of the defaults in the constructor
+        self.db_visit_modules = db_visit_modules or self.db_visit_modules
+        self.blueprints = blueprints or self.blueprints
+        self.template_folder = template_folder or self.template_folder
+        if app:
+            # Not really intended to be used this way, but it fits the flask extension paradigm
+            # and could conceivably be set up in lieu of KEG_REGISTERED_COMPONENTS
+            self.init_app(app)
+
+    def set_dotted_path(self, dotted_path):
+        # The Keg app sets this when following KEG_REGISTERED_COMPONENTS
+        self.relative_dotted_path = dotted_path

```

Comment:
Once the `relative_dotted_path` is set, can it be set again? 

If not, we should add a `None` check and raise an exception if it is being set again. That would require that we use `None` as the default value.

Reply:
I went a different direction here. It's possible that two apps in the same space would share the same component, which would necessarily require the init sequence for the component to be run twice. I can't think of a scenario where the path to the component would be different between the two apps, but we'll just have it passed on `init_app` rather than storing it on the instance.